### PR TITLE
45 ny hp rates assign number of cars to households

### PR DESCRIPTION
**Pull request by alxsmith**
Some notes on the predictive fit
0 vehicles: 952 correctly predicted, but 15,454 predicted as 1, 8,011 predicted as 2
1 vehicle: 25,674 correctly predicted, but 2,546 predicted as 0, 12,628 predicted as 2
2+ vehicles: 36,143 correctly predicted, but 980 predicted as 0, 19,453 predicted as 1

**Key Insights:**
Overall Accuracy: 51.5% - Not great, but reasonable for a 3-class problem

**Class Imbalance Issues:**
The model struggles with 0 vehicles (only 4% recall)
Does better with 1 vehicle (63% recall)
Best with 2+ vehicles (64% recall)

**Prediction Bias:**
Model tends to overpredict 1 vehicle (most common class)
Underpredicts 0 vehicles significantly - _in our original model, we had a binary for NYC. it feels weird to implement this for a general model but may be a factor here_
Slightly underpredicts 2+ vehicles

### CODE REVIEW

**Comment by jpvelez** [comment]
- 2 EVs max is fine
- 0 should def be an option, let user code handle the 0 case.
- Income assignment: that's fine as a first pass, but this should ultimately use our inferred `income` variable that we will make available via `bsf`.
- add `sklearn` as a dev dependency and if a tool doesn't like it and the tool is wrong, silence the too.
- let's go with basic mode for now and make a ticket for making a better model later.

File: buildstock_fetch/utils/ev_utils.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        3: ["IL", "IN", "MI", "OH", "WI"],  # East North Central
+        4: ["IA", "KS", "MN", "MO", "NE", "ND", "SD"],  # West North Central
+        5: ["DE", "DC", "FL", "GA", "MD", "NC", "SC", "VA", "WV"],  # South Atlantic
+        6: ["AL", "KY", "MS", "TN"],  # East South Central
+        7: ["AR", "LA", "OK", "TX"],  # West South Central
+        8: ["AZ", "CO", "ID", "MT", "NV", "NM", "UT", "WY"],  # Mountain
+        9: ["AK", "CA", "HI", "OR", "WA"],  # Pacific
+    }
+
+    for division, states in census_divisions.items():
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:50]: This whole function should be dictionary that maps state abbreviations to census division numbers. Symbol tables were invented to avoid these kinds of `for` loops! We're trying to have a civilization here. ;)

File: buildstock_fetch/utils/ev_utils.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+def load_metadata(metadata_path: str) -> pl.DataFrame:
+    """
+    Load and parse the ResStock metadata parquet file.
+
+    Args:
+        metadata_path: Path to the metadata parquet file
+
+    Returns:
+        DataFrame with columns including 'bldg_id', 'occupants', 'income', 'metro', 'puma.
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:90]: typo: `puma`

File: buildstock_fetch/utils/ev_utils.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    metadata_df = metadata_df.with_columns([
+        pl.when(pl.col("occupants") == "10+").then(pl.lit("10")).otherwise(pl.col("occupants")).alias("occupants_temp")
+    ])
+
+    # Now cast to numeric
+    metadata_df = metadata_df.with_columns([pl.col("occupants_temp").cast(pl.Int64).alias("occupants")]).drop(
+        "occupants_temp"
+    )
+
+    # # Create household size bins
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:121]: Fix or delete?

File: buildstock_fetch/utils/ev_utils.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        .then(pl.lit(None))
+        .otherwise(pl.col("income"))
+        .alias("income")
+    ])
+
+    # Convert income ranges to numeric midpoints
+    metadata_df = metadata_df.with_columns([pl.col("income").map_elements(assign_income_midpoints).alias("income")])
+
+    # Extract last 5 characters from PUMA
+    metadata_df = metadata_df.with_columns([pl.col("puma").str.slice(-5).alias("puma")])
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:147]: @mshron what do we think about all this variable recycling? Isn't that smell that suggests this should be one big pipe?

File: buildstock_fetch/utils/ev_utils.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+    Returns:
+        DataFrame with trip records filtered for the specified state's census division
+    """
+    if not Path(nhts_path).exists():
+        msg = f"NHTS file not found: {nhts_path}. Please run `just download-nhts` to download the data."
+        raise FileNotFoundError(msg)
+
+    # Define the columns we need
+    needed_columns = [
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:168]: love this

File: buildstock_fetch/utils/ev_utils.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    # Get the census division for this state
+    state_division = get_census_division_for_state(state)
+
+    # Filter to only keep census division for this state
+    nhts_df = nhts_df.filter(pl.col("CENSUS_D") == state_division)
+
+    # Remove the CENSUS_D column since we don't need it anymore
+    nhts_df = nhts_df.drop("CENSUS_D")
+
+    nhts_df = nhts_df.rename({
```
Comment (jpvelez) [inline buildstock_fetch/utils/ev_utils.py:193]: Maybe define this dictionary when `needed_columns` is now, keeping your nice comments, and extract `needed_columns` with it?

File: buildstock_fetch/ev_demand.py [hunk 7]
```diff
@@ -79,18 +105,31 @@ class EVDemandCalculator:
     7. Assign battery capacities
     """
 
-    def __init__(self, metadata_path: Optional[str] = None):
+    def __init__(
+        self,
+        metadata_df: pl.DataFrame,
+        nhts_df: pl.DataFrame,
+        pums_df: pl.DataFrame,
+        weather_df: Optional[pl.DataFrame] = None,
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:113]: Actually optionL

File: buildstock_fetch/ev_demand.py [hunk 8]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
             -2.0659e-10,  # a_5 (quintic term)
         ])
 
-    def load_metadata(self, metadata_path: Optional[str] = None) -> pl.DataFrame:
+    def fit_vehicle_ownership_model(self, pums_df: pl.DataFrame) -> Any:
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:148]: no type signature for this badboy?

File: buildstock_fetch/ev_demand.py [hunk 9]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
+        Fit a multinomial logistic regression model to predict number of vehicles per household using PUMS data.
+        Results limited to 0, 1, or 2 vehicles.
 
         Args:
-            metadata_path: Path to the metadata parquet file. If None, uses self.metadata_path
+            pums_df: DataFrame with PUMS household data, including 'occupants', 'income', 'metro', 'vehicles'
 
         Returns:
-            DataFrame with columns including 'building_id', 'occupants', 'inc', 'metro_area' (PUMA), etc.
+            Trained model object
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:157]: for sklearn?

File: buildstock_fetch/ev_demand.py [hunk 10]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
         """
-        path = metadata_path or self.metadata_path
-        if path is None:
-            raise MetadataPathError()
+        # Preprocess data: replace vehicles > 2 with 2, drop nulls, encode categorical
+        pums_df = pums_df.with_columns(
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:160]: I'd move feature engineering to a separate method, so that this one just takes whatever features we've got and does the training, and the feature engineering method can evolve as we add features.

File: buildstock_fetch/ev_demand.py [hunk 11]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
+
+        # Encode target and scale features
+        self.target_encoder = LabelEncoder()
+        y_encoded = self.target_encoder.fit_transform(y.get_column("vehicles").to_numpy())
+
+        self.scaler = StandardScaler()
+        X_scaled = self.scaler.fit_transform(X_encoded.to_numpy())
+
+        # Fit model with sample weights
+        self.vehicle_ownership_model = LogisticRegression(solver="lbfgs", max_iter=1000, random_state=self.random_state)
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:194]: party like its 2011

File: buildstock_fetch/ev_demand.py [hunk 12]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
 
-        Returns:
-            DataFrame with trip records, including trip weights, household features, and trip parameters
-        """
-        # TODO: Implement NHTS data loading and preprocessing
-        # This is a placeholder - replace with actual NHTS loading logic
-        self.nhts_df = pl.DataFrame()  # Placeholder
-        return self.nhts_df
+        # Step 2: Encode categorical variables
+        features_encoded = features.clone()
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:233]: I wonder if any of the feature engineering logic currently in the fit method could be reused (once feature engineering is moved to its own method) rather than reimplemented?

File: buildstock_fetch/ev_demand.py [hunk 13]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
 
-        Returns:
-            Trained model object
-        """
-        # TODO: Implement multinomial logistic regression or similar
-        # This is a placeholder - replace with actual model fitting
-        self.vehicle_ownership_model = "placeholder_model"
         return self.vehicle_ownership_model
 
     def predict_num_vehicles(self, metadata_df: Optional[pl.DataFrame] = None) -> pl.DataFrame:
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:199]: Looks like `metadata_df` isn't actually optional?

File: buildstock_fetch/ev_demand.py [hunk 14]
```diff
@@ -106,89 +145,108 @@ def __init__(self, metadata_path: Optional[str] = None):
...
+        # Step 5: Decode predictions and add to DataFrame
+        predictions_decoded = self.target_encoder.inverse_transform(predictions_encoded)
+
+        # Add predictions to original DataFrame
+        result_df = df.with_columns(pl.Series(predictions_decoded).alias("vehicles"))
+
+        return result_df
 
     def sample_vehicle_profiles(
         self, metadata_df: Optional[pl.DataFrame] = None
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:252]: ditto on `metadata_df` not actually being Optional?

File: buildstock_fetch/ev_demand.py [hunk 15]
```diff
@@ -356,142 +414,80 @@ def assign_battery_capacity(self, daily_kwh: pl.Series) -> pl.Series:
...
+    #         Dictionary containing all output DataFrames and Series
+    #     """
+    #     # Step 1: Load metadata
+    #     metadata_df = self.load_metadata()
 
-        # Step 2: Load PUMS data and fit vehicle ownership model
-        pums_df = pl.read_csv(pums_path)  # TODO: Adjust based on actual PUMS format
-        self.fit_vehicle_ownership_model(pums_df)
+    #     # Step 2: Load PUMS data and fit vehicle ownership model
+    #     pums_df = pl.read_csv(pums_path)  # TODO: Adjust based on actual PUMS format
```
Comment (jpvelez) [inline buildstock_fetch/ev_demand.py:436]: Delete all this commented out stuff?

File: tests/test_ev_demand.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+    EVDemandCalculator,
+    MetadataDataFrameError,
+    VehicleOwnershipModelError,
+)
+
+
+@pytest.fixture
+def sample_pums_data():
+    """Create sample PUMS data for testing."""
+    return pl.DataFrame({
```
Comment (jpvelez) [inline tests/test_ev_demand.py:15]: good stuff

File: utils/ev_data/inputs/download_nhts_data.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
```
Comment (jpvelez) [inline utils/ev_data/inputs/download_nhts_data.py:1]: good stuff

**Review by jpvelez — CHANGES_REQUESTED** [review]
Just a bunch of nits.

File: tests/test_ev_demand.py [hunk 18]
```diff
@@ -0,0 +1,262 @@
+import numpy as np
```
Comment (jpvelez) [inline tests/test_ev_demand.py:1]: @alxsmith dumb question: where is the code the generates the accuracy metrics for the models? Maybe that should be a notebook or a script?

Comment (jpvelez) [reply tests/test_ev_demand.py:1]: a script in `utils/`

File: buildstock_fetch/ev_demand.py (hunk 7 above)
Comment (jpvelez) [reply buildstock_fetch/ev_demand.py:113]: ok

File: buildstock_fetch/utils/ev_utils.py [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+
+    return pums_df
+
+
+def load_weather_data(weather_path: str) -> pl.DataFrame:
+    """
+    Load hourly weather data (e.g., temperature) for a given location.
+
+    Args:
+        weather_path: Path to the weather data file (e.g., TMY3 CSV or EPW)
```
[the remark this answers is not in the document]
Comment (alxsmith) [reply buildstock_fetch/utils/ev_utils.py:249]: tbd

File: buildstock_fetch/utils/ev_utils.py [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+
+    Returns:
+        DataFrame with at least columns ['datetime', 'temperature']
+    """
+    # if not Path(weather_path).exists():
+    #     msg = f"Weather file not found: {weather_path}"
+    #     raise FileNotFoundError(msg)
+
+    # TODO: Implement weather data loading
```
[the remark this answers is not in the document]
Comment (alxsmith) [reply buildstock_fetch/utils/ev_utils.py:258]: no, it is used for converting miles driven to kwh. data not yet available https://github.com/switchbox-data/buildstock-fetch/issues/48

File: buildstock_fetch/ev_demand.py (hunk 12 above)
Comment (alxsmith) [reply buildstock_fetch/ev_demand.py:233]: good thought, i think it should be will try and recycle

File: buildstock_fetch/utils/ev_utils.py [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+
+def load_pums_data(pums_path: str, metadata_path: str) -> pl.DataFrame:
+    """
+    Load and preprocess the PUMS data for a specific state.
+
+    Args:
+        pums_path: Path to the PUMS data file
+        metadata_path: Path to metadata file for metro-PUMA mapping
+    """
+    if not Path(pums_path).exists():
```
[the remark this answers is not in the document]
Comment (mshron) [reply buildstock_fetch/utils/ev_utils.py:216]: i.e. make a function called `_check_pums_present(path)`

File: buildstock_fetch/utils/ev_utils.py [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+        "VEHCASEID": "vehicle_id",
+        "STRTTIME": "start_time",
+        "ENDTIME": "end_time",
+        "TRPMILES": "miles_driven",
+    })
+
+    return nhts_df
+
+
+def load_pums_data(pums_path: str, metadata_path: str) -> pl.DataFrame:
```
[the remark this answers is not in the document]
Comment (mshron) [reply buildstock_fetch/utils/ev_utils.py:208]: It says

> `# If metadata_path is provided, join with metro-PUMA mapping`

and then does a check, so presumably you can run this without `metadata_path`?

File: buildstock_fetch/ev_demand.py (hunk 11 above)
Comment (mshron) [reply buildstock_fetch/ev_demand.py:194]: I think he's just making a joke about logistic regressions feeling quaint in a world of trillion-parameter deep learning models

File: tests/test_ev_demand.py [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+        "metro": ["urban", "urban", "suburban", "suburban"],
+        "vehicles": [0, 1, 2, 2],
+        "hh_weight": [2.0, 1.5, 1.0, 0.5],  # Different weights
+    })
+
+    model = calculator.fit_vehicle_ownership_model(pums_data)
+
+    # Check that model was fitted successfully
+    assert model is not None
+    assert calculator.vehicle_ownership_model is not None
```
[the remark this answers is not in the document]
Comment (mshron) [reply tests/test_ev_demand.py:110]: If you wanted to, you could also do a quick logistic regression in R to test that we're setting things up correctly with sklearn, but might be overkill

**Comment by mshron** [comment]
Re: carrying around all of `sklearn`... all else equal if we literally just need multinomial logistic regression I would use `statsmodels`. It provides a lot more out-of-the-box validation and quality feedback than `sklearn` does for generalized regression models.

Having said that, you've already done the work, so not really worth going back and changing it, but just something to think about for next time.

File: buildstock_fetch/ev_demand.py [hunk 24]
```diff
@@ -356,142 +414,80 @@ def assign_battery_capacity(self, daily_kwh: pl.Series) -> pl.Series:
...
-        }
+    #     return {
+    #         "metadata_with_vehicles": metadata_with_vehicles,
+    #         "trip_schedules": trip_schedules,
+    #         "battery_capacities": battery_capacities,
+    #     }
 
 
-# Example usage
+# # Example usage
```
[the remark this answers is not in the document]
Comment (alxsmith) [reply buildstock_fetch/ev_demand.py:474]: this is not the final run script. its just a way for me to check all the components along the way during dev
